lib/ansible/plugins/inventory/tower.py

```diff
@@ -0,0 +1,201 @@
+    read_all_configs_from_env_var = False
+
+    def read_tower_inventory(self, tower_host, tower_user, tower_pass, inventory, license_type, verify_ssl=True):
+        if not re.match('(?:http|https)://', tower_host):
+            tower_host = "https://{tower_host}".format(tower_host=tower_host)
+        inventory_id = inventory.replace('/', '')
+        inventory_url = "/api/v2/inventories/{inv_id}/script/?hostvars=1&towervars=1&all=1".format(inv_id=inventory_id)
+        inventory_url = urljoin(tower_host, inventory_url)
+        config_url = urljoin(tower_host, "/api/v2/config/")
+        try:
+            if license_type != "open":
+                config_response = requests.get(config_url,
+                                               auth=HTTPBasicAuth(tower_user, tower_pass),
+                                               verify=verify_ssl)
+                if config_response.ok:
+                    source_type = config_response.json()['license_info']['license_type']
+                    if not source_type == license_type:
+                        raise AnsibleParserError("Tower server licenses must match: source: {source} local: {local}".
+                                                 format(source=source_type, local=license_type))
+                else:
+                    raise AnsibleParserError("Failed to validate the license of the remote Tower: {response}".
+                                             format(response=config_response))
+
+            response = requests.get(inventory_url,
+                                    auth=HTTPBasicAuth(tower_user, tower_pass),
+                                    verify=verify_ssl)
+            if not response.ok:
+                # If the GET /api/v2/inventories/N/script is not HTTP 200, print the error code
+                raise AnsibleParserError("Connection to remote host failed: {response}".format(response=response))
+            try:
+                # Attempt to parse JSON
+                return response.json()
+            except (ValueError, TypeError) as e:
+                # If the JSON parse fails, print the ValueError
+                raise AnsibleParserError("Failed to parse json from host: {err}".format(err=e))
+        except requests.ConnectionError as e:
+            raise AnsibleParserError("Connection to remote host failed: {err}".format(err=e))
+
+    def verify_file(self, path):
+        if path.endswith('tower_inventory'):
+            self.read_all_configs_from_env_var = True
+            return True
+        elif super(InventoryModule, self).verify_file(path):
+            return path.endswith('tower_inventory.yml') or path.endswith('tower_inventory.yaml')
+        else:
+            return False
+
+    def get_option(self, option, hostvars=None):
+        '''
+            The default get_option cannot cast int to string, even if the type is set.
+            We override that behavior and force all numbers into string.
+        '''
```

Comment:
we should fix the `ensure_type` function to do this, instead of working around it in every plugin

Reply:
👍

Reply:
this still has not changed, what is the actual issue you hit to require this 'type enforcement'?